book-cw/part-2-coding/regex.adoc

```diff
@@ -1,3 +1,348 @@
+If you've used it before but never taken the time to learn its full potential, this chapter should be helpful as well.
+
+Seasoned regex coders can keep me honest.
+This book is open source; help me bring the power of this intimidating but mercifully straightforward scripting syntax to struggling documentarians everywhere.
+
+.What-ex?
+[NOTE]
+I pronounce the shortened term _REJ-ex_, though I believe I've heard more developers use _REGG-ex_.
+Or maybe it just strikes me more starkly when I hear it pronounced differently than I say it.
+
+== What is Regex?
+
+A regular expression is a way of describing a pattern using common characters in order to search for the described pattern in another body of characters.
+The target can be pretty much any form of text, such as a conventional document or a flattened dataset.
+A regex parsing engine uses these expressions to evaluate target texts for matches.
+
+Put differently by the regex-focused site Rexegg, “A regex is a text string that describes a pattern that a regex engine uses in order to find text (or positions) in a body of text, typically for the purposes of validating, finding, replacing, or splitting.”
+
+At its simplest application, think of regex as a much smarter version of the `*` or `%` wildcard symbols you've probably used in other contexts.
+Where a conventional wildcard finds _any and all text_, regex can find shockingly detailed patterns without false positives.
+
+The string `codewriting.org` is matched by the following regex patterns:
+
+* `[a-zA-Z0-9][a-zA-Z0-9\-]{1,61}[a-zA-Z0-9]\.[a-zA-Z]{2,}`
+* `.*@\.[a-z]+`
+* `codewriting\.org`
+
+You'll note that all of these are more complicated than the non-regex search term:
+
+* `codewriting.org`
+
+Probably in most cases of a find-and-replace task, I don't get to use regex.
+
+But what if we needed to find all instances of `http://` protocol indicator for a `codewriting.org` URL in our docs and replace `http:` with `https:`.
+We don't want this for every instance of `http:` in the document, but we also cannot count on the domain string to always appear intact, as `http://codewriting.org`, which we could easily replace with `https://codewriting.org`.
+Making matters more complicated, numerous instances of a `codewriting.org` URL appear with a hostname or “subdomain”, such as `docs.codewriting.org`.
+To make sure we caught them all manually, we'd first have to create a list, and then iterate through it one by one replacing the whole string.
+
+Here's the simple regex pattern that will save us all this trouble:
+
+[source,regex]
+----
+http\:\/\/([a-z]+\.)?codewriting\.org
+----
+
+This regex would efficiently match the following strings:
+
+* `http://codewriting.org`
+* `http://git.codewriting.org`
+* `http://staging.codewriting.org`
+
+The single set of parentheses in the regex will capture any “group” it matches and store it as a variable we can later reference to reinsert the captured string.
```

Comment:
Need a better summary of the regex code here. Explain more clearly what's within the parens.